Add more actor types in Core to ForSyDeIR

SamuelMiksits opened the pull request:
Adding as draft as I am still debugging a couple of things.

## Changes
Adds the remaining actors to Core -> ForSyDe IR.

I do this by adding cases for them in:

- `translateCoreExpr` - Add the other actors
- `getActorSplit` - Add the other actors
- `translateSystemExpr` - Add so that it supports 3,4 inputs as well (covers all other actors). NOTE: I additionally reversed the `arguments` lists, see below:

## Tests

I am running tests on `models/examples/SDF_example_006.hs` and `models/examples/SDF_example_007.hs`. Here is a drawing of both of them:
(sry for only having drawing and no tikz)
<img width="1336" height="1124" alt="sdf_006_007" src="https://github.com/user-attachments/assets/ac212cab-0ab3-4624-9f3a-b2e8b75ad0fb" />

### `SDF_example_006.hs`

Generally this looks correct. But here is why I reversed the arguments in the `translateSystemExpr`, without doing so I received the following output:

```
IRSystem(
  {"s_in"}, {"s_out"},
  {
    IRActor("a_a", Actor14, "f_1", {"s_in"}, {"s_1", "s_2", "s_3", "s_4"}),
    IRActor("a_b", Actor21, "f_2", {"s_1", "s_2"}, {"s_5"}),
    IRActor("a_c", Actor12, "f_3", {"s_3"}, {"s_6", "s_7"}),
    IRActor("a_d", Actor21, "f_4", {"s_5", "s_6"}, {"s_8"}),
    IRActor("a_e", Actor31, "f_5", {"s_8", "s_7", "s_4"}, {"s_out"})
  },
  {
    IRSignal("s_in", ("s_in", 1), ("a_a", 2)),
    IRSignal("s_2", ("a_a", 3), ("a_b", 1)),
    IRSignal("s_1", ("a_a", 1), ("a_b", 3)),
    IRSignal("s_3", ("a_a", 2), ("a_c", 2)),
    IRSignal("s_6", ("a_c", 2), ("a_d", 2)),
    IRSignal("s_5", ("a_b", 1), ("a_d", 2)),
    IRSignal("s_4", ("a_a", 4), ("a_e", 2)),
    IRSignal("s_7", ("a_c", 1), ("a_e", 1)),
    IRSignal("s_8", ("a_d", 1), ("a_e", 4)),
    IRSignal("s_out", ("a_e", 1), ("s_out", 1))
  },
```

the issue here is that `s_1` goes from a->b with output rate 1 from a and input rate 3 on b, and `s_2` goes from a->b with output rate 3 from a and input rate 1 on b. This is clearly wrong, showing that they are going to the "wrong inputs" (despite `a_b` having arguments `{"s_1", "s_2"}`, but if I reverse the arguments I get the following graph:

```
IRSystem(
  {"s_in"}, {"s_out"},
  {
    IRActor("a_a", Actor14, "f_1", {"s_in"}, {"s_1", "s_2", "s_3", "s_4"}),
    IRActor("a_b", Actor21, "f_2", {"s_2", "s_1"}, {"s_5"}),
    IRActor("a_c", Actor12, "f_3", {"s_3"}, {"s_6", "s_7"}),
    IRActor("a_d", Actor21, "f_4", {"s_6", "s_5"}, {"s_8"}),
    IRActor("a_e", Actor31, "f_5", {"s_4", "s_7", "s_8"}, {"s_out"})
  },
  {
    IRSignal("s_in", ("s_in", 1), ("a_a", 2)),
    IRSignal("s_1", ("a_a", 1), ("a_b", 1)),
    IRSignal("s_2", ("a_a", 3), ("a_b", 3)),
    IRSignal("s_3", ("a_a", 2), ("a_c", 2)),
    IRSignal("s_5", ("a_b", 1), ("a_d", 2)),
    IRSignal("s_6", ("a_c", 2), ("a_d", 2)),
    IRSignal("s_8", ("a_d", 1), ("a_e", 2)),
    IRSignal("s_7", ("a_c", 1), ("a_e", 1)),
    IRSignal("s_4", ("a_a", 4), ("a_e", 4)),
    IRSignal("s_out", ("a_e", 1), ("s_out", 1))
  },
```
where the input order is now reversed on 
```
IRActor("a_b", Actor21, "f_2", {"s_2", "s_1"}, {"s_5"}),
```
but all of the signals are correctly routed (i.e s_1 goes from 1->1 and s_2 goes from 3->3). Is the issue actually that we are applying `(Var a1)` etc in the wrong order or something else? I also don't understand why I did not have this issue in `SDF_example_007.hs`

### `SDF_example_007.hs`

I get the following ForSyDe IR:

```
IRSystem(
  {"s_in_a", "s_in_b"}, {"s_out"},
  {
    IRActor("a_a", Actor12, "f_1", {"s_in_a"}, {"s_1", "s_2"}),
    IRActor("a_b", Actor13, "f_2", {"s_in_b"}, {"s_3", "s_4", "s_5"}),
    IRActor("a_c", Actor32, "f_3", {"s_4", "s_3", "s_2"}, {"s_6", "s_7"}),
    IRActor("a_d", Actor41, "f_4", {"s_5", "s_7", "s_6", "s_1"}, {"s_out"})
  },
  {
    IRSignal("s_in_b", ("s_in_b", 1), ("a_b", 1)),
    IRSignal("s_in_a", ("s_in_a", 1), ("a_a", 2)),
    IRSignal("s_2", ("a_a", 1), ("a_c", 2)),
    IRSignal("s_3", ("a_b", 2), ("a_c", 4)),
    IRSignal("s_4", ("a_b", 1), ("a_c", 2)),
    IRSignal("s_1", ("a_a", 2), ("a_d", 4)),
    IRSignal("s_6", ("a_c", 2), ("a_d", 2)),
    IRSignal("s_7", ("a_c", 2), ("a_d", 2)),
    IRSignal("s_5", ("a_b", 2), ("a_d", 4)),
    IRSignal("s_out", ("a_d", 1), ("s_out", 1))
  },
```
the only thing that looks wrong is 
```
    IRSignal("s_out", ("a_d", 1), ("s_out", 1))
```
where the correct output should be
```
    IRSignal("s_out", ("a_d", 2), ("s_out", 1))
```
Is it something about having an actor with multiple outputs causing other models to be correct (binder to output), and since this one is from actor to output it gets 1 as default output rate?  `SDF_example_002.hs` has correct output rate 2, but I am unsure if the other models with single output only have 1 as token output rate and as a result are "accidentally correct".

SamuelMiksits committed:
Add more actor types in Core to ForSyDeIR

SamuelMiksits commented on src/CoreIRToForSyDeIR.hs:
```diff
@@ -221,7 +221,26 @@ translateSystemExpr initialContext expr = case expr of
         a1Id = showPpr (flags initialContext) a1
         a2Id = showPpr (flags initialContext) a2
-        arguments = [(a2Id), (a1Id)]
+        arguments = [(a1Id), (a2Id)]
```
reversed list order, see PR description

SamuelMiksits commented on src/CoreIRToForSyDeIR.hs:
```diff
@@ -382,8 +461,20 @@ getActorSplit :: ActorType -> Int
+  Actor41 -> 4
+  Actor42 -> 4
+  Actor43 -> 4
+  Actor44 -> 4
```
I get 
```
src/CoreIRToForSyDeIR.hs:478:3: warning: [GHC-53633] [-Woverlapping-patterns]
    Pattern match is redundant
    In a case alternative: _ -> ...
    |
478 |   _ -> error "getActorSplit: unsupported actor type"
    |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
since we have pattern matched on (exhausted) every single possible combination inside the Actor type. We still want to keep this right since we could add Actor types from other ForSyDe MoCs?

sthaeron replied:
Decided to delete the error pattern matching as future MoC models will probably still be using the types defined by `ActorType` since they are not specific to SDF models.

sthaeron commented:
> Is it something about having an actor with multiple outputs causing other models to be correct (binder to output), and since this one is from actor to output it gets 1 as default output rate? `SDF_example_002.hs` has correct output rate 2, but I am unsure if the other models with single output only have 1 as token output rate and as a result are "accidentally correct".

Yes you're right if the actor only has one output its associated `Binder` will be a `PcId` rather than a `Binding`. And in the case of a `PcId` the output rate defaults to 1 rather than properly checking the output rates and getting the rate at index zero.

sthaeron commented:
> * NOTE: I additionally reversed the `arguments` lists, see below:

This change makes the order of input arguments for SDF example 2 wrong. I also noticed that before your reversing of the arguments list the target rates for a few signals was wrong. I think the solution is not reversing the arguments list be rather that there is a problem with the `createSignals` function than needs fixing.

SamuelMiksits commented:
> This change makes the order of input arguments for SDF example 2 wrong. I also noticed that before your reversing of the arguments list the target rates for a few signals was wrong. I think the solution is not reversing the arguments list be rather that there is a problem with the `createSignals` function than needs fixing.

Okay yeah I agree it is probably the wrong solution since the output list is in the "correct" order and the input list being reversed just looks wrong anyways. It is possible to just change from
```hs
arguments = [(a1Id), (a2Id), (a3Id), (a4Id)]
```
to
```hs
arguments = reverse [(a1Id), (a2Id), (a3Id), (a4Id)]
```
if you quickly need to toggle between

sthaeron committed:
fix: refactored and fixed getting outputs from actors with more than 1 output

sthaeron commented:
@SamuelMiksits I rebased and fixed the issues with the new core translation.
- Fixed the outputs order issue by reversing arguments and reversing input rates when creating signals from arguments.
- Updated the `getSourceFromArgument` function to return the the output rate at index zero when argument directly represents a process constructor.
- Did some refactoring updating function names and comments.

SamuelMiksits commented:
> @SamuelMiksits I rebased and fixed the issues with the new core translation.
> 
> * Fixed the outputs order issue by reversing arguments and reversing input rates when creating signals from arguments.
> * Updated the `getSourceFromArgument` function to return the the output rate at index zero when argument directly represents a process constructor.
> * Did some refactoring updating function names and comments.

Thank you! I will take a look at it now

sthaeron reviewed (approved):
After fixing the problems with the new core translation I checked the output from your new SDF examples. Seems to me that your new pattern matching for creating and applying actors are working. Also I checked my fixes are working by double checking the signals output using your diagrams. From my end it works great and is ready to merge!

SamuelMiksits commented:
>After fixing the problems with the new core translation I checked the output from your new SDF examples. Seems to me that your new pattern matching for creating and applying actors are working. Also I checked my fixes are working by double checking the signals output using your diagrams. From my end it works great and is ready to merge!

Nice, I used the LSP and the visual output matched the graph as well lol

sthaeron committed:
refactor: found a method to simplify actor pattern matching

sthaeron commented:
@SamuelMiksits I found a way of simplifying the pattern matching of actors, what do you think?

SamuelMiksits commented:
> @SamuelMiksits I found a way of simplifying the pattern matching of actors, what do you think?

Wow yeah that is an improvement. Some lines are very long (like 160 cols) but if we actually want to do something about it it should probably be in ormolu anyways. Otherwise unless @MichelD-Abo has something to say about it you can merge
